src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/EventMap.cs

```diff
@@ -6,13 +6,14 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Linq;
+using System.Threading;
 using Microsoft.CodeAnalysis.ErrorReporting;
 
 namespace Roslyn.Utilities
 {
     internal class EventMap
     {
-        private readonly NonReentrantLock _guard = new();
+        private readonly SemaphoreSlim _guard = new(initialCount: 1);
```

Comment:
💡 This one doesn't use cancellation tokens, but it does include specific assertions about holding a lock. It seems like a good candidate for `object`/`Monitor`.